### PR TITLE
Adding support for body option

### PR DESCRIPTION
**Pull request by himanshu226**
Adding support for Body Option Storage.

### CODE REVIEW

File: lib/requester/core-body-builder.js [hunk 1]
```diff
@@ -116,9 +126,27 @@ formDataBodyReducer = function (data, param) {
 module.exports = {
     /**
      * @param {Object} content - request body content
+     * @param {Object} request - request object
      * @returns {Object}
      */
-    raw: function (content) {
+    raw: function (content, request) {
+        var mode = _.get(request, 'body.mode'),
```
Comment (codenirvana) [inline lib/requester/core-body-builder.js]: its `raw` mode body-builder, no need to check mode again.

File: lib/requester/core-body-builder.js [hunk 2]
```diff
@@ -116,9 +126,27 @@ formDataBodyReducer = function (data, param) {
...
+            options = _.get(request, 'body.options'),
```
Comment (codenirvana) [inline lib/requester/core-body-builder.js]:
```suggestion
            options = _.get(request, 'body.options.raw') || {},
```

File: lib/requester/core-body-builder.js [hunk 3]
```diff
@@ -116,9 +126,27 @@ formDataBodyReducer = function (data, param) {
...
+            contentType,
+            contentLanguage;
+
+        // Add `Content-Type` header from body options if not set already
+        if (options && !request.headers.has(CONTENT_TYPE_HEADER.key)) {
+            contentType = _.get(options[mode], 'contentType');
```
Comment (codenirvana) [inline lib/requester/core-body-builder.js]: the option name is not `contentType`

Comment (himanshu226) [reply lib/requester/core-body-builder.js]: This is according to spec.

File: lib/requester/core-body-builder.js [hunk 4]
```diff
@@ -28,6 +28,22 @@ var _ = require('lodash'),
...
 
+    /**
+     * Use to map content-type from body options
+     *
+     * @private
+     * @type {Object}
+     */
+    CONTENT_TYPE = {
+        'text-html': 'text/html',
+        'text': 'text',
```
Comment (codenirvana) [inline lib/requester/core-body-builder.js]: No `Content-Type` for language `text`.

File: lib/requester/core-body-builder.js [hunk 5]
```diff
@@ -116,9 +132,23 @@ formDataBodyReducer = function (data, param) {
 module.exports = {
     /**
      * @param {Object} content - request body content
+     * @param {Object} request - request object
```
Comment (codenirvana) [inline lib/requester/core-body-builder.js]:
```suggestion
     * @param {Request} [request] - request object
```

File: lib/requester/core-body-builder.js [hunk 6]
```diff
@@ -116,9 +132,23 @@ formDataBodyReducer = function (data, param) {
...
      * @returns {Object}
      */
-    raw: function (content) {
+    raw: function (content, request) {
+        var options = _.get(request, 'body.options.raw'),
```
Comment (codenirvana) [inline lib/requester/core-body-builder.js]:
```suggestion
        var contentLanguage = _.get(request, 'body.options.raw.language'),
```

File: test/integration/file-uploads/request-body.test.js [hunk 7]
```diff
@@ -117,11 +117,10 @@ describe('file upload in request body', function () {
 
             var resp = JSON.parse(testrun.response.getCall(3).args[2].stream.toString());
 
-            expect(resp).to.have.property('data', '{\n\t"key1":"value1",\n\t"key2": 2\n}\n');
+            expect(resp.data).to.be.eql({key1: 'value1', key2: 2});
             expect(resp).to.nested.include({
                 'headers.content-length': '33'
             });
-            expect(resp.headers['content-type']).to.equal('text/plain');
```
Comment (codenirvana) [inline test/integration/file-uploads/request-body.test.js]: Don't remove this assertion. Instead, add an assertion for missing/inferred `Content-Type` from the file type.

File: test/integration/file-uploads/request-body.test.js [hunk 8]
```diff
@@ -293,11 +292,9 @@ describe('file upload in request body', function () {
                 });
                 expect(req.body.file).to.have.property('content');
 
-                expect(resp).to.deep.nested.include({
-                    'headers.content-length': '33',
-                    data: '{\n\t"key1":"value1",\n\t"key2": 2\n}\n'
-                });
-                expect(resp.headers['content-type']).to.equal('text/plain');
```
Comment (codenirvana) [inline test/integration/file-uploads/request-body.test.js]: Add inferred `Content-Type`.

File: test/integration/mutations/request.test.js [hunk 9]
```diff
@@ -281,7 +281,7 @@ describe('request mutations', function () {
             expect(request.body).to.deep.include({mode: 'raw', raw: 'postman'});
 
             expect(response).to.have.property('code', 200);
-            expect(responseBody).to.have.property('data', 'postman');
+            expect(responseBody.data).to.be.eql({});
```
Comment (codenirvana) [inline test/integration/mutations/request.test.js]: this seems wrong.

File: test/integration/request-body/raw.test.js [hunk 10]
```diff
@@ -41,7 +41,7 @@ describe('Request Body Mode: raw', function () {
                 responseBody = JSON.parse(response.stream.toString());
 
             expect(response).to.have.property('code', 200);
-            expect(responseBody).to.have.property('data', 'POSTMAN');
+            expect(responseBody.data).to.be.eql({});
```
Comment (codenirvana) [inline test/integration/request-body/raw.test.js]: Why there's no `data`? Is `Content-Type` required for postman-echo?

File: test/unit/requester-core.test.js [hunk 11]
```diff
@@ -446,6 +445,81 @@ describe('requester util', function () {
...
+                        mode: 'raw',
+                        options: {
+                            raw: {
+                                language: 'application-json'
+                            }
+                        }
+                    }
+                });
+
+                expect(request.headers.has('Content-Type')).to.be.false;
```
Comment (codenirvana) [inline test/unit/requester-core.test.js]: Assert presence of `body.options`

File: test/unit/requester-core.test.js [hunk 12]
```diff
@@ -446,6 +445,81 @@ describe('requester util', function () {
...
+                        raw: '{"beta":"bar"}',
+                        options: {
+                            raw: {
+                                language: 'application-json'
+                            }
+                        }
+                    }
+                });
+
+                expect(requesterCore.getRequestBody(request)).to.eql({
```
Comment (codenirvana) [inline test/unit/requester-core.test.js]: before this assert missing content-type.

File: test/integration/request-body/raw.test.js (hunk 10 above)
Comment (himanshu226) [reply test/integration/request-body/raw.test.js]: Yes, there is no data when we remove content-type.

File: test/integration/file-uploads/request-body.test.js (hunk 7 above)
Comment (himanshu226) [reply test/integration/file-uploads/request-body.test.js]: We have content-type as  '' (empty string). Should I add expect for empty string?

**Pull request merged**